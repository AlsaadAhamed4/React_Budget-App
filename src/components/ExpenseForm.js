import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from "react-dates";
import 'react-dates/initialize';

const now = moment();

console.log(now.format('MMM - Do - YYYY'));

export default class ExpenseForm extends React.Component {

    constructor(props) {
        super(props);

        this.state = {
            description: props.editData ? props.editData.description : '',
            amount: props.editData ? (props.editData.amount / 100).toString() : '',
            note: props.editData ? props.editData.note : '',
            createdAt: props.editData ? moment(props.editData.createdAt) : moment(),
            calendarFocused: false,
            errMsg: '',
            buttonName :  props.editData ? 'Edit Expense' : 'Add Expense'
        }
    }

    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    }

    onAmountChange = (e) => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) { //only numbers and 2 decimal numbers 
            this.setState(() => ({ amount }));
        }
    }

    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    }

    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(() => ({ createdAt }));
        }
    }

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }));
    }

    onSubmit = (e) => {
        e.preventDefault();
        if (!this.state.description && !this.state.amount) {
            this.setState(() => ({ errMsg: 'Please Specify Description & Amount' }));
        }
        else if (!this.state.description) {
            this.setState(() => ({ errMsg: 'Please Specify Description' }));
        }
        else if (!this.state.amount) {
            this.setState(() => ({ errMsg: 'Please Specify Amount' }));
        }
        else {
            this.setState(() => ({ errMsg: '' }));
            //since we use this form for edit page as well so we cannot dispatch a action here. We use a prop on call to be used by both edit and create
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    }

    render() {
        return (
            <div>
                {this.state.errMsg && <p>{this.state.errMsg}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type='text' placeholder='Description' autoFocus value={this.state.description} onChange={this.onDescriptionChange} />
                    <input type='text' placeholder='Amount' value={this.state.amount} onChange={this.onAmountChange} />
                    <SingleDatePicker date={this.state.createdAt} onDateChange={this.onDateChange} focused={this.state.calendarFocused} onFocusChange={this.onFocusChange} numberOfMonths={1} isOutsideRange={() => false} />
                    <textarea placeholder='Add a note for your expense (optional)' value={this.state.note} onChange={this.onNoteChange} />
                    <button>{this.state.buttonName}</button>
                </form>
            </div>
        )
    }
}

//single date picker to be able display all past days we use isOutsideRange={()=> false}
//to be able to display some days then we say isOutsideRange={(days)=> logic}