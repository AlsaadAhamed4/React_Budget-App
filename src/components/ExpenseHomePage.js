import React from 'react';


const ExpenseHomePage = () => (
    <div>
        <p>This is expense page</p>
    </div>
);

export default  ExpenseHomePage;
