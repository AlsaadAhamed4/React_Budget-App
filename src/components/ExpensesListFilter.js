import React from 'react';
import { connect } from 'react-redux';
import {setTextFilterAction} from '../Actions/FiltersAction';


//this component is responsible for displaying input feild with value of applied filter

//here if we dont use onchange then we cant change the text value
//To set the state to specified filter we need to dispatch it store , we can do it by calling props.dispatch() in onchange
//here in line 15 i am able to write the state by using dispatch and a action 
const ExpensesListFilter = (props) => (
    <div>
    <br/>
        <label>Filter Used is:</label><input type='text' value={props.filters.text} onChange={(e)=>{
            props.dispatch(setTextFilterAction(e.target.value))
        }} />
    </div>
);


const mapSateToProps = (state) => {
    return {
        filters: state.filters
    }
};

export default connect(mapSateToProps)(ExpensesListFilter);