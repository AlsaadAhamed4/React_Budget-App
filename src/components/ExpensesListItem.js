import React from 'react';
import { removeExpenseAction, editExpenseAction } from '../Actions/ExpenseAction';
import { NavLink } from 'react-router-dom';

const ExpensesListItem = ({ description, amount, createdAt, id, dispatch }) => (
    <div>
        <NavLink to={`/edit/${id}`}><h1>{description}</h1></NavLink>
        <p>{amount}</p>
        <p>{createdAt}</p>
    </div >
);

export default ExpensesListItem;