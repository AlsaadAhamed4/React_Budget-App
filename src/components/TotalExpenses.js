import React from 'react';
import { connect } from 'react-redux';
import { totalExpenses } from '../Selectors/Total-Expenses';

export const TotalExpenses = (props) => (
    <div>
        {props.total && <p>Viewing {props.total.totalExpenseCount} expenses - Totaling {(props.total.totalExpenseAmount)}</p>}
    </div>
);

const mapStateToprops = (state) => (
    {
        total: totalExpenses(state.expenses)
    }
);

export default connect(mapStateToprops)(TotalExpenses);