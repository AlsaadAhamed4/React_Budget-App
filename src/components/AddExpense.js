import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpenseAction } from '../Actions/ExpenseAction';

//made a class base function to remove inline function for submit by using mapDispatchToProps
//export for testing

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.onSubmit(expense) //dispatching through mapDispatchToProps
        this.props.history.push('/')
    }
    render(){
        return(
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm onSubmit={this.onSubmit} />
        </div>
        )
    }
}


// const AddExpensePage = (props) => {
//     console.log(props);
//     return (
//         <div>
//             <h1>Add Expense</h1>
//             <ExpenseForm onSubmit={(expense) => {
//                 props.dispatch(addExpenseAction(expense))
//                 props.history.push('/')
//             }} />
//         </div>
//     )
// };

//passing a props to expenseform so that we could use it for edit 
//we are passing a props function so that we recive submitted data an then we dispatch it to store.

//to dispatch we only need to connect it and add it to store

//push is used to redirect to a path

//note : We can make props.dispatch(addExpenseAction(expense)) to be simple , As connect has two args
// 1 mapStateToProps  2 mapDispacthToProps  in dispatch we can pass it so that testing can be easier

const mapDispatchToProps = (dispatch) => (
    {
        onSubmit: (expense) => dispatch(addExpenseAction(expense))
    }
);

export default connect(undefined, mapDispatchToProps)(AddExpensePage);