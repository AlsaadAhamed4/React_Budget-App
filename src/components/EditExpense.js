import React from 'react';
import ExpenseForm from "../components/ExpenseForm";
import { connect } from 'react-redux';
import { editExpenseAction, removeExpenseAction } from '../Actions/ExpenseAction';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    OnRemove = (e) => {
        this.props.removeExpense(this.props.expense.id);
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <p>Edit Expense</p>
                <ExpenseForm editData={this.props.expense} onSubmit={this.onSubmit} />
                <button onClick={this.OnRemove}>Remove</button>
            </div>
        )
    }
}


/* const EditExpensePage = (props) => {
    console.log(props.expense);
    return (
        <div>
            <p>Edit Expense {props.match.params.id}</p>
            <ExpenseForm editData={props.expense} onSubmit={(expense) => {
                props.dispatch(editExpenseAction(props.match.params.id, expense))
                props.history.push('/')
            }} />
            <button onClick={(e) => {
                props.dispatch(removeExpenseAction({ id: props.expense.id }))
                props.history.push('/')
            }}>Remove</button>
        </div>
    )
}; */

const mapStateToProp = (state, props) => {
    return {
        expense: state.expenses.find(({ id }) => id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch, props) => (
    {
        editExpense: (id, expense) => dispatch(editExpenseAction(id, expense)),
        removeExpense: (id) => dispatch(removeExpenseAction({ id }))
    }
);

export default connect(mapStateToProp, mapDispatchToProps)(EditExpensePage);
