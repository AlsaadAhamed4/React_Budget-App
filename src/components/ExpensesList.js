//this js is passed to ExpenseHome.js to render the view in UI

import React from 'react';
import { connect } from 'react-redux'; //used to connect to store and its a HOC 
import ExpensesListItem from './ExpensesListItem';
import filterExpenses from '../Selectors/FilterExpenses';


const ExpensesList = (props) => (
    <div>
        <h1>Expenses List</h1>
        {props.expenses.length}
        {props.expenses.map((exp) => {
            return <ExpensesListItem {...exp} key={exp.id} />
        })}
    </div>
);

//connect syntax connect(function with state as args)(Name of Component to connect)
//connect function
//standard name used as mapStateToProps
const mapStateToProps = (state) => {
    return {
        expenses: filterExpenses(state.expenses,state.filters) //this can be accessed as props in connected component in my case 'ExpenseList'
    };
};

export default connect(mapStateToProps)(ExpensesList); //connecting redux store to react component
