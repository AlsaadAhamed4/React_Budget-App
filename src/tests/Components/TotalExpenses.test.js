import React from 'react';
import { shallow } from 'enzyme';
import { TotalExpenses } from '../../components/TotalExpenses';
import numeral from 'numeral'

let wrapper, totalExpensesCount, totalExpensesAmount;

test('should render totalExpenses to be empty', () => {
    wrapper = shallow(<TotalExpenses totalExpensesCount={0} totalExpensesAmount={numeral(0).format('$0,0.00')} />);
    expect(wrapper).toMatchSnapshot();
})


beforeEach(() => {

    totalExpensesCount = 1;
    totalExpensesAmount = numeral(222.10).format('$0,0.00');

    wrapper = shallow(<TotalExpenses totalExpensesCount={totalExpensesCount} totalExpensesAmount={totalExpensesAmount} />);
});

test('should render totalExpenses', () => {
    expect(wrapper).toMatchSnapshot();
});
