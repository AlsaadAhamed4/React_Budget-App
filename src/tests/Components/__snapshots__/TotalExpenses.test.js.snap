// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render totalExpenses 1`] = `
<div
  className="totalexpense-header"
>
  <div
    className="content-container"
  >
    <h1
      className="totalexpense-content__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       - Totaling 
      <span>
        $222.10
      </span>
    </h1>
    <div
      className="totalexpense-header__action"
    >
      <[object Object]
        className="button"
        to="/create"
      >
        Add Expense
      </[object Object]>
    </div>
  </div>
</div>
`;

exports[`should render totalExpenses to be empty 1`] = `
<div
  className="totalexpense-header"
>
  <div
    className="content-container"
  >
    <h1
      className="totalexpense-content__title"
    >
      Viewing 
      <span>
        0
      </span>
       
      expenses
       - Totaling 
      <span>
        $0.00
      </span>
    </h1>
    <div
      className="totalexpense-header__action"
    >
      <[object Object]
        className="button"
        to="/create"
      >
        Add Expense
      </[object Object]>
    </div>
  </div>
</div>
`;
